core/src/sumeragi/mod.rs

```diff
@@ -53,42 +59,209 @@ trait Consensus {
+        wsv_guard.metrics.connected_peers.set(online_peers_count);
+        for domain in domains {
+            wsv_guard
+                .metrics
+                .accounts
+                .get_metric_with_label_values(&[domain.id().name.as_ref()])
+                .wrap_err("Failed to compose domains")?
+                .set(domain.accounts().len() as u64);
+        }
+        Ok(())
+    }
+
+    pub fn latest_block_hash_for_use_by_block_sync(&self) -> HashOf<VersionedCommittedBlock> {
+        self.internal
+            .latest_block_hash_for_use_by_block_sync
+            .lock()
+            .unwrap()
+            .clone()
+    }
+
+    pub fn blocks_after_hash(
+        &self,
+        block_hash: HashOf<VersionedCommittedBlock>,
+    ) -> Vec<VersionedCommittedBlock> {
+        self.internal
+            .wsv_for_public_use
+            .lock()
+            .unwrap()
+            .blocks_after_hash(block_hash)
+    }
+
+    pub fn blocks_from_height(&self, block_height: usize) -> Vec<VersionedCommittedBlock> {
+        self.internal
+            .wsv_for_public_use
+            .lock()
+            .unwrap()
+            .blocks_from_height(block_height)
+    }
+
+    pub fn get_random_peer_for_block_sync(&self) -> Option<Peer> {
+        use rand::{prelude::SliceRandom, RngCore, SeedableRng};
+
+        let rng = &mut rand::rngs::StdRng::from_entropy();
+        let peers = self
+            .internal
+            .current_online_peers
+            .lock()
+            .expect("lock on online peers for get random peer")
+            .iter()
+            .map(|peer| Peer::new((*peer).clone()))
+            .collect::<Vec<Peer>>();
+        if peers.is_empty() {
+            return None;
+        } else {
+            let mut sorted_peers = peers;
+            sorted_peers.sort();
+            Some(sorted_peers[rng.next_u32() as usize % sorted_peers.len()].clone())
+        }
+    }
+
+    pub fn wsv_mutex_access(&self) -> std::sync::MutexGuard<WorldStateView> {
+        self.internal.wsv_for_public_use.lock().unwrap()
+    }
+
+    pub fn initialize_and_start_thread(
+        sumeragi: Arc<Self>,
+        latest_block_hash: HashOf<VersionedCommittedBlock>,
+        latest_block_height: u64,
+    ) {
+        std::thread::Builder::new()
+            .name("sumeragi thread".to_string())
+            .spawn(move || {
+                fault::run_sumeragi_main_loop(
+                    &sumeragi.internal,
+                    latest_block_hash,
+                    latest_block_height,
+                );
+                info!("Sumeragi Thread has shut down.");
+            });
+    }
+
+    pub fn stop_thread(&self) {
+        self.internal
+            .sumeragi_state_machine_data
+            .lock()
+            .expect("lock to stop sumeragi thread")
+            .sumeragi_thread_should_exit = true;
+    }
+
+    pub fn update_online_peers(&self, online_peers: Vec<PeerId>) {
```

Comment:
At some point we could change this to patch update approach, not to replace whole vector every time.